tokio/src/task/mod.rs

```diff
@@ -363,49 +381,74 @@ cfg_rt_core! {
         pub(crate) fn header(&self) -> &Header {
             self.raw.header()
         }
+    }
 
-        pub(crate) fn into_raw(self) -> NonNull<Header> {
-            let raw = self.raw.into_raw();
-            mem::forget(self);
-            raw
+    impl<S: 'static> Notified<S> {
+        pub(crate) unsafe fn from_raw(ptr: NonNull<Header>) -> Notified<S> {
+            Notified(Task::from_raw(ptr))
+        }
+
+        pub(crate) fn header(&self) -> &Header {
+            self.0.header()
         }
     }
 
-    impl<S: Schedule> Task<S> {
-        /// Returns `self` when the task needs to be immediately re-scheduled
-        pub(crate) fn run<F>(self, mut executor: F) -> Option<Self>
-        where
-            F: FnMut() -> Option<NonNull<S>>,
-        {
-            if unsafe {
-                self.raw
-                    .poll(&mut || executor().map(|ptr| ptr.cast::<()>()))
-            } {
-                Some(self)
-            } else {
-                // Cleaning up the `Task` instance is done from within the poll
-                // function.
+    cfg_rt_threaded! {
+        impl<S: 'static> Task<S> {
+            pub(crate) fn into_raw(self) -> NonNull<Header> {
+                let ret = self.header().into();
                 mem::forget(self);
-                None
+                ret
+            }
+        }
+
+        impl<S: 'static> Notified<S> {
+            pub(crate) fn into_raw(self) -> NonNull<Header> {
+                self.0.into_raw()
             }
         }
+    }
 
+    impl<S: Schedule> Task<S> {
         /// Pre-emptively cancel the task as part of the shutdown process.
         pub(crate) fn shutdown(self) {
-            self.raw.cancel_from_queue();
+            self.raw.shutdown();
+            mem::forget(self);
```

Comment:
All these `mem::forget` make me even more sure that `Task` should usually be `TaskRef`, which could then internally hold a `ManuallyDrop<RawTask>` or some such.

Reply:
`Task` is pretty much a "task ref" right now. Either way, shutdown was cleaned up some w.r.t ref counting.